File: lib/src/common/auth/OSSFederationCredentialProvider.java
package com.alibaba.sdk.android.oss.common.auth;

import com.alibaba.sdk.android.oss.OSSClientException;
import com.alibaba.sdk.android.oss.common.OSSLog;
import com.alibaba.sdk.android.oss.common.utils.DateUtil;

/**
 * Created by zhouzhuo on 11/4/15.
 */
 abstract class OSSFederationCredentialProvider implements OSSCredentialProvider {

     volatile OSSFederationToken cachedToken;

    /**
     * Gets the valid STS token. The subclass needs to implement this function.
     *
     * @return The valid STS Token
     */
     abstract OSSFederationToken getFederationToken() throws OSSClientException;

     synchronized OSSFederationToken getValidFederationToken() throws OSSClientException {
        // Checks if the STS token is expired. To avoid returning staled data, here we pre-fetch the token 5 minutes a head of the real expiration.
        // The minimal expiration time is 15 minutes
        if (cachedToken == null
                || DateUtil.getFixedSkewedTimeMillis() / 1000 > cachedToken.getExpiration() - 5 * 60) {

            if (cachedToken != null) {
                OSSLog.logDebug("token expired! current time: " + DateUtil.getFixedSkewedTimeMillis() / 1000 + " token expired: " + cachedToken.getExpiration());
            }
            cachedToken = getFederationToken();
        }

        return cachedToken;
    }

     OSSFederationToken getCachedToken() {
        return cachedToken;
    }
}

File: lib/src/common/auth/oss_federation_credential_provider.dart
import 'package:aliyun_oss_dart_sdk/src/common/oss_log.dart';
import 'package:aliyun_oss_dart_sdk/src/common/utils/date_util.dart';

import 'oss_credential_provider.dart';
import 'oss_federation_token.dart';

abstract class OSSFederationCredentialProvider
    implements OSSCredentialProvider {
  OSSFederationToken? _cachedToken;

  Future<OSSFederationToken?> getValidFederationToken() async {
    // Checks if the STS token is expired. To avoid returning staled data, here we pre-fetch the token 5 minutes a head of the real expiration.
    // The minimal expiration time is 15 minutes
    final nowSecond = DateUtil.getFixedSkewedTimeMillis() / 1000;
    if (_cachedToken == null || nowSecond > _cachedToken!.expiration - 5 * 60) {
      if (_cachedToken != null) {
        OSSLog.logDebug(
            "token expired! current time: $nowSecond token expired: ${_cachedToken!.expiration}");
      }
      _cachedToken = await getFederationToken();
    }

    return _cachedToken;
  }

  OSSFederationToken? get cachedToken {
    return _cachedToken;
  }
}
